### PR TITLE
feat(elements): Add `<Link />` component

### PR DESCRIPTION
**Pull request by alexcarpenter**
## Description

This PR introduces a new Elements `<Link />` that ties into the internal ClerkRouter that preserves search params when moving between sign-in and sign-up flows.

### Usage

```tsx
import * as Clerk from '@clerk/elements/common'
import NextLink from 'next/link'

function SignInPage() {
  return <>
    <Clerk.Link navigate="sign-up">Sign up</Clerk.Link>

    <Clerk.Link navigate="sign-up">
      {url => <NextLink href={url}>Sign up</NextLink>}
    </Clerk.Link>
  </>
}
```

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/elements/src/react/common/link.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+const paths: Record<Destination, keyof Pick<ClerkOptions, 'signInUrl' | 'signUpUrl'>> = {
+  'sign-in': 'signInUrl',
+  'sign-up': 'signUpUrl',
+};
+
+export function Link({ navigate, children, ...rest }: LinkProps) {
+  const router = useClerkRouter();
+  const clerk = useClerk();
+  const destiationUrl = router.makeDestinationUrlWithPreservedQueryParameters(
```
Comment (LauraBeatris) [inline packages/elements/src/react/common/link.tsx:20]:
```suggestion
  const destinationUrl = router.makeDestinationUrlWithPreservedQueryParameters(
```

**Review by panteliselef — APPROVED** [review]

File: packages/types/src/clerk.ts [hunk 2]
```diff
@@ -104,6 +104,8 @@ export interface Clerk {
    */
   loaded: boolean;
 
+  __internal_getOption<K extends keyof ClerkOptions>(key: K): ClerkOptions[K];
```
Comment (LauraBeatris) [inline packages/types/src/clerk.ts:107]: I'd suggest to place this as a helper outside of the main `Clerk` interface as we're using this mostly for typing purposes.

**Review by LauraBeatris — APPROVED** [review]

File: packages/elements/src/react/common/link.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { useClerk } from '@clerk/shared/react';
+import { useClerkRouter } from '@clerk/shared/router';
+import type { ClerkOptions } from '@clerk/types';
+import React from 'react';
+
+type Destination = 'sign-in' | 'sign-up';
+export interface LinkProps extends Omit<React.HTMLAttributes<HTMLAnchorElement>, 'children'> {
+  navigate: Destination;
+  children: React.ReactNode | ((url: string) => React.ReactNode);
```
Comment (panteliselef) [inline packages/elements/src/react/common/link.tsx:9]: Can we do { url } in case we need to add more ?

Comment (alexcarpenter) [reply packages/elements/src/react/common/link.tsx:9]: I am fine with that. Do we keep the prop name as `url`?

File: packages/types/src/clerk.ts (hunk 2 above)
Comment (alexcarpenter) [reply packages/types/src/clerk.ts:107]: @LauraBeatris can you share more? I am not completely following the change requested.

File: packages/elements/src/react/common/link.tsx (hunk 3 above)
Comment (alexcarpenter) [reply packages/elements/src/react/common/link.tsx]: updated ✅

File: packages/types/src/clerk.ts (hunk 2 above)
Comment (LauraBeatris) [reply packages/types/src/clerk.ts:107]: Ohhh, I'm sorry! Reading the code at first, I thought it was only an internal utility method for accessing a property against `ClerkOptions` keys - you can ignore that one.

**Comment by alexcarpenter** [comment]
!snapshot

**Comment by clerk-cookie** [comment]
Hey @alexcarpenter - the snapshot version command generated the following package versions:
| Package | Version |
| --- | --- |
| @clerk/astro | 1.4.3-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/backend | 1.15.3-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/chrome-extension | 1.3.25-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/clerk-js | 5.30.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/elements | 0.18.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/clerk-expo | 2.2.31-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/expo-passkeys | 1.0.0-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/express | 1.3.5-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/fastify | 2.0.7-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/localizations | 3.4.2-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/nextjs | 6.1.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/clerk-react | 5.14.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/remix | 4.2.43-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/clerk-sdk-node | 5.0.56-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/shared | 2.11.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/tanstack-start | 0.4.19-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/testing | 1.3.17-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/themes | 2.1.41-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/types | 4.29.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |
| @clerk/ui | 0.1.12-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 |

Tip: Use the snippet copy button below to quickly install the required packages.
`@clerk/astro`
```sh
npm i @clerk/astro@1.4.3-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/backend`
```sh
npm i @clerk/backend@1.15.3-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/chrome-extension`
```sh
npm i @clerk/chrome-extension@1.3.25-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/clerk-js`
```sh
npm i @clerk/clerk-js@5.30.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/elements`
```sh
npm i @clerk/elements@0.18.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/clerk-expo`
```sh
npm i @clerk/clerk-expo@2.2.31-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/expo-passkeys`
```sh
npm i @clerk/expo-passkeys@1.0.0-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/express`
```sh
npm i @clerk/express@1.3.5-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/fastify`
```sh
npm i @clerk/fastify@2.0.7-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/localizations`
```sh
npm i @clerk/localizations@3.4.2-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/nextjs`
```sh
npm i @clerk/nextjs@6.1.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/clerk-react`
```sh
npm i @clerk/clerk-react@5.14.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/remix`
```sh
npm i @clerk/remix@4.2.43-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/clerk-sdk-node`
```sh
npm i @clerk/clerk-sdk-node@5.0.56-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/shared`
```sh
npm i @clerk/shared@2.11.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/tanstack-start`
```sh
npm i @clerk/tanstack-start@0.4.19-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/testing`
```sh
npm i @clerk/testing@1.3.17-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/themes`
```sh
npm i @clerk/themes@2.1.41-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/types`
```sh
npm i @clerk/types@4.29.1-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```
`@clerk/ui`
```sh
npm i @clerk/ui@0.1.12-snapshot.v409c69b9b3f950967f11d0ca682eb536aa406684 --save-exact
```

**Comment by alexcarpenter** [comment]
!snapshot

**Pull request merged**